llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp

```diff
@@ -51,6 +53,7 @@ RealtimeSanitizerPass::RealtimeSanitizerPass(
 PreservedAnalyses RealtimeSanitizerPass::run(Function &F,
                                              AnalysisManager<Function> &AM) {
   if (F.hasFnAttribute(Attribute::SanitizeRealtime)) {
+    assert(!F.hasFnAttribute(Attribute::NoSanitizeRealtime));
```

Comment:
Why do we need NoSanitizeRealtime, when Address and Memory sanitizers can exist without them?

Reply:
The main reason this is necessary is we need to have a way to turn off the rtsan stack for all function calls below the function with the nosanitize attribute. 

If you mark a function as needing the real-time sanitizer, all functions invoked by that function are subject to it's constraints (no allocations, no locks, no system calls). This is enforced via a stack, which is popped or pushed via __rtsan_realtime_enter and __rtsan_realtime_exit, inserted in this file.

When you mark a function as not needing to be inspected by the realtime sanitizer, it must change the stack such so these constraints are no longer in place. The nosanitize function AS WELL AS functions invoked by it can allocate, take locks, or call system calls.

Address and Memory sanitizer don't have this hierarchical model. 

To make the example concrete...
Realtime sanitizer

```cpp
int ChildWithViolation() {
  int* v = new int[100]; // SHOULD BE VIOLATION, can't alloc in nonblocking context
}

__attribute__((no_sanitize("realtime")))
int Parent() {
  return ChildWithViolation(); // does not throw an error, because we are within a no_sanitize("realtime") context
}

int main() [[clang::nonblocking]]{
  Parent();
}
```

Marking parent with no_sanitize("realtime") disables the realtime sanitizer for this function and all functions this function calls.

Address sanitizer, does not work like this, if you try to do something similar:

```cpp
int ChildWithViolation() {
  int* v = new int[100];
  return v[100];
}

__attribute__((no_sanitize("address")))
int Parent() {
  return ChildWithViolation();
}

int main() {
  Parent();

 	return 0;
}
```

```sh
> ./build/helloWorld
=================================================================
==24510==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x000107915bd0 at pc 0x000104ea3e14 bp 0x00016af5e950 sp 0x00016af5e948
READ of size 4 at 0x000107915bd0 thread T0
    #0 0x104ea3e10 in ChildWithViolation() main.cpp:17
    #1 0x104ea3e34 in Parent() main.cpp:22
    #2 0x104ea3e58 in main main.cpp:26
    #3 0x1958960dc  (<unknown module>)
...
```

The nosanitize attribute only turns it off for that particular function, not "down the stack"

Reply:
Why it needs to be attribute?

`__lsan::ScopedDisabler` achieves the same  and even can be more precise, but limiting to a code block.

Reply:
Thanks for pointing that out, I wasn't aware it existed!

**Pros of ScopedDisabler:**

- Enables finer-grained disabling within a function.
- Avoids modifying LLVM attributes or Clang, simplifying the implementation.
- Precedent exists with lsan.

**Cons of ScopedDisabler:**

- Doesn't follow existing patterns for disabling interceptors, unlike asan and others.
- Users need to recreate build system infrastructure for conditional compilation with `-fsanitize=realtime`.

The biggest concern is that second "con" -- users would need to manually include out-of-project headers and manage build conditions, leading to some potential issues:

1. Duplication of effort across users.
2. Difficulty in locating the required header files. (and conditionally including them if the sanitizer is enabled, leading to more complex build systems)
3. Risk of users re-implementing ScopedDisabler in their codebases.

The `nosanitize` attribute is comparatively simpler, which "just works" without extra steps. I would prefer whatever solution we go forward with, it uses a construct that would be as non-invasive as the nosanitize approach to avoid burdening users with additional setup. Are there other solutions that provide this level of ease to the end user?

For finer-grained interception, we've advised extracting code into new functions and applying the nosanitize attribute. This approach is a little more invasive but clearly indicates that the code is treated differently.

Reply:
> Thanks for pointing that out, I wasn't aware it existed!
> 
> **Pros of ScopedDisabler:**
> 
> * Enables finer-grained disabling within a function.
> * Avoids modifying LLVM attributes or Clang, simplifying the implementation.
> * Precedent exists with lsan.
> 
> **Cons of ScopedDisabler:**
> 
> * Doesn't follow existing patterns for disabling interceptors, unlike asan and others.
> * Users need to recreate build system infrastructure for conditional compilation with `-fsanitize=realtime`.
> 
> The biggest concern is that second "con" -- users would need to manually include out-of-project headers and manage build conditions, leading to some potential issues:

conditions should not be a problem if you check __feature in rtsan head.
So it could be included unconditionally, and just compile into no-op without RTSan
we should do the same for all sanitizers, we have some inconsistency there.

> 
> 1. Duplication of effort across users.

Code modification is needed. It should be either attribute or C++ object either way, so I don't see much difference in duplication.

> 2. Difficulty in locating the required header files. (and conditionally including them if the sanitizer is enabled, leading to more complex build systems)

Should be unconditional. Header installed with clang.
However with g++ it may need a condition.

But I don't think it's a big issue. We used to preprocessor in C++.

> 3. Risk of users re-implementing ScopedDisabler in their codebases.

Why it's a problem? They can do so even you go with existing attribute. I am 100% sure that ScopedDisabler like will be needed, as function scope may affect to much.

Also inliner and other passes can be affected by attribute. Regular C++ code will be more predictable for end user.

> 
> The `nosanitize` attribute is comparatively simpler, which "just works" without extra steps. I would prefer whatever solution we go forward with, it uses a construct that would be as non-invasive as the nosanitize approach to avoid burdening users with additional setup. Are there other solutions that provide this level of ease to the end user?

lsan has larger user base then rtsan, at least for now, and it was never an issue. Attribute approach can work for lsan as well.

> 
> For finer-grained interception, we've advised extracting code into new functions and applying the nosanitize attribute. This approach is a little more invasive but clearly indicates that the code is treated differently.

I am not aware of context for this, can you point me there? How attribute is better here then ScopedDisabler?


So unless we have strong reason, I would prefer we solve the same problem in all component consistently.

Reply:
> conditions should not be a problem if you check __feature in rtsan head.
So it could be included unconditionally, and just compile into no-op without RTSan
we should do the same for all sanitizers, we have some inconsistency there.

Could you elaborate on this, perhaps with a small code example? I think if this were possible my concern would be alleviated!

Reply:
(actually, I think I have it figured out, going to test this for a little bit and get back to you. Thanks for the guidance)

Reply:
This works great for us @vitalybuka ! Thanks for explaining this to me. You will see some follow up reviews today that accomplish this goal.

Closing this review.

Reply:
That review is posted here: https://github.com/llvm/llvm-project/pull/106736